### PR TITLE
[Serve][Doc] Add Batching User Guide

### LINKED ISSUE #27445: [serve][docs] `Request Batching` user guide

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
Add a new page discussing how to use the batching decorator.

## Related issue number
Closes #27445

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:3]: Serve offers a request batching feature that can improve your overall service throughput by sacrificing some latency.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:5]: This guide teaches you how to:

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:8]: - use Serve's `@serve.batch` decorator
- configure `@serve.batch` decorator

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:9]: Machine Learning (ML) frameworks like Tensorflow, PyTorch, and Scikit-Learn efficiently   support and evaluate multiple samples at the same time.

OR drop the adverb: both are in active voice and strong action verbs.

Machine Learning (ML) frameworks like Tensorflow, PyTorch, and Scikit-Learn support and evaluate multiple samples at the same time.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:11]: When a request arrives, Serve puts the request in a queue. This queue buffers requests to form a batch sample. The batch is then be picked up by the model for evaluation. After the evaluation, the result batch is split up , and each response is sent individually.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:15]: Let's take a look at a simple example by modifying the `MyModel` class to accept a batch.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:23]: Make the following changes to the batch decorator method signature:

- The class method is declared as an async method because the decorator batches in asyncio event loop.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:24]: The class method accepts a list of its original input types as input. For example, `arg1: int, arg2: str` should be changed to `arg1: List[int], arg2: List[str]`.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:25]: The class method  returns a list. The length of the return list and the input list must be of equal lengths for the decorator to split the output evenly and return a corresponding response back to its respective request.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:36]: You can supply two optional parameter to the decorators.

File: doc/source/serve/batching-guide.md
Comment (dmatrix) [inline doc/source/serve/batching-guide.md:46]: "In this scenario, the second model will always evaluate twice when the first model finishes a full batch."
Don't follow what you mean. Why twice? may be explain a bit.

**Review by dmatrix — COMMENTED** [review]
made minor nits.

**Comment by edoakes** [comment]
Will do a pass after @dmatrix's comments are addressed

File: doc/source/serve/batching-guide.md
Comment (architkulkarni) [inline doc/source/serve/batching-guide.md:36]:
```suggestion
You can supply two optional parameters to the decorators.
- `batch_wait_timeout_s` controls how long Serve should wait for a batch once the first request arrives.
```

File: doc/source/serve/batching-guide.md
Comment (architkulkarni) [inline doc/source/serve/batching-guide.md:42]:
```suggestion
`max_batch_size` ideally should be a power of 2 (2, 4, 8, 16, ...) because CPUs and GPUs are both optimized for data of these shapes. Large batch sizes incur a high memory cost as well as latency penalty for the first few requests.
```

Comment (architkulkarni) [inline doc/source/serve/batching-guide.md:42]:
```suggestion
`max_batch_size` ideally should be a power of 2 (2, 4, 8, 16, ...) because CPUs and GPUs are both optimized for data of these shapes. Large batch sizes incur a high memory cost as well as latency penalty for the first few requests.
```

File: doc/source/serve/batching-guide.md
Comment (architkulkarni) [inline doc/source/serve/batching-guide.md:44]:
```suggestion
`batch_wait_timeout_s` should be set considering the end to end latency SLO (Service Level Objective). After all, the first requests could potentially wait this long for a full batch, adding to its latency cost. For example, if your latency target is 150ms, and the model takes 100ms to evaluate the batch, the `batch_wait_timeout_s` should be set to a value much lower than 150ms - 100ms = 50ms.
```

Comment (architkulkarni) [reply doc/source/serve/batching-guide.md:44]: Just to make it totally clear where 50ms comes from

File: doc/source/serve/batching-guide.md
Comment (architkulkarni) [inline doc/source/serve/batching-guide.md:46]:
```suggestion
When using batching in a Serve Deployment Graph, the relationship between an upstream node and a downstream node might affect the performance as well. Consider a chain of two models where the first model sets `max_batch_size=8` and second model sets `max_batch_size=6`. In this scenario, when the first model finishes a full batch or 8, the second model will finish one batch of 6 and then wait to fill the next batch, which will initially only be partially filled with 8 - 6 = 2 requests, incurring latency cost. The batch size of downstream models should ideally be multiples or divisors of the upstream models to ensure the batches play well together.
```

**Review by architkulkarni — COMMENTED** [review]
Added some wording suggestions.  As part of this PR, can we also delete other existing batching docs or shorten them and link to this one?

File: doc/source/serve/batching-guide.md
Comment (edoakes) [inline doc/source/serve/batching-guide.md:1]:
```suggestion
# Request Batching
```

File: doc/source/serve/batching-guide.md
Comment (edoakes) [reply doc/source/serve/batching-guide.md:3]: I think the intention is "without sacrificing latency" ?

Comment (edoakes) [reply doc/source/serve/batching-guide.md:3]: @simon-mo please add a few sentences describing why batching is important for ML models (efficient vectorized computation)

Comment (architkulkarni) [reply doc/source/serve/batching-guide.md:3]: It might be subtle, the worst-case latency of a request will increase because it'll be waiting for a batch instead of being processed instantly.  But if the overall throughput increases, the average latency would decrease because there won't be as long of a queue in front of each replica.  Maybe we just omit the latency part from this first sentence

**Comment by simon-mo** [comment]
Comments addressed! I think we can keep the tutorial example which uses PyTorch and show case how to compile batches into tensors and others.

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@simon-mo you need codeowner approval from someone here

**Review by dmatrix — APPROVED** [review]

**Pull request merged**
